### PR TITLE
Add is_reference API

### PR DESCRIPTION
**Pull request by Robbybp**
## Summary/Motivation:
Several users have noticed that there is no officially supported way to distinguish a component that was created by the `Reference` function. However, it is very useful to be able to make this distinction. Also, in some cases, it is useful to get the object ~~(slice)~~ from which a reference was created. 

This PR addresses the first point by adding an ~~`is_reference`~~ `owns_data` method to `PyomoObject`, and the second by tacking on a `referent` attribute to the component returned by the `Reference` function. 

## Changes proposed in this PR:
- Add `owns_data` method to `PyomoObject` and `IndexedComponent`
- Add `referent` attribute to component returned by `Reference`
- Use `owns_data` to address a hiccup in `IndexedComponent` that causes iteration over references-to-references to be incredibly slow. 

One intent of this PR is to provide a concrete talking point for the proposals in #1640.

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

**Comment by Robbybp** [comment]
As @jsiirola has pointed out, `owns_data` might be a better name than `is_reference` as the former is more general. In particular, if we ever want to decouple references and slices, we could end up with something like a `SliceComponent` function that creates something distinct from a "reference" that doesn't own its own data.

I've changed the method name here to reflect this and have added some tests.

**Comment by michaelbynum** [comment]
I like this a lot.

**Comment by Robbybp** [comment]
@carldlaird @michaelbynum @andrewlee94 We are trying to come up with the name for the method to test if an object owns its own data. Any suggestions on an `is_*` method that accomplishes this? Our best idea for now is `is_reference`, but this may not be general enough.

**Review by blnicho — APPROVED** [review]
One minor question, otherwise this looks fine.

File: pyomo/core/base/indexed_component.py [hunk 1]
```diff
@@ -258,6 +258,10 @@ def is_indexed(self):
         """Return true if this component is indexed"""
         return self._index is not UnindexedComponent_set
 
+    def is_reference(self):
+        """ Return True if this component is a reference. """
```
[the remark this answers is not in the document]
Comment (Robbybp) [reply pyomo/core/base/indexed_component.py]: I think that's a good idea. Have updated.

**Review by jsiirola — APPROVED** [review]

**Pull request merged**